app/assets/javascripts/save-on-change.coffee

```diff
@@ -179,21 +179,29 @@ class @ForwardBlocker
     @binding.bind 'navigate_away', (e,opts) =>
       @navigate_away(opts.click_element, opts.action_to_perform)
 
-  prevent_forward_navigation_for: (blocker) ->
-    if blocker not in @blockers
-      @blockers.push blocker
+  get_blocker_index: (blocker) ->
+    firstBlocker = (b for b in @blockers when b.blocker is blocker)[0]
+    @blockers.indexOf(firstBlocker)
+
+  get_latest_disabled_forward_nav_message: ->
+    (b.disabledForwardNavMessage for b in @blockers when b.disabledForwardNavMessage?).reverse()[0]
+
+  prevent_forward_navigation_for: (blocker, disabledForwardNavMessage) ->
+    if @get_blocker_index(blocker) is -1
+      @blockers.push({blocker, disabledForwardNavMessage})
       @update_display()
 
   enable_forward_navigation_for: (blocker) ->
-    index = @blockers.indexOf(blocker)
+    index = @get_blocker_index(blocker)
     if index > -1
       @blockers.splice(index,1)
       @update_display()
 
   navigate_away: (click_element, action_to_perform) ->
     if @block_for_element(click_element)
       $('.question').addClass('did_try_to_navigate')
-      modalDialog(false, t('PLEASE_SUBMIT'))
+      message = @get_latest_disabled_forward_nav_message() or t('PLEASE_SUBMIT')
```

Comment:
Instead of this, why not have all blockers have a message. This way the question objects would just pass in `t('PLEASE_SUBMIT')` as their message. Then the 'get_latest...._message' could be changed to just get the latest blocker object and the message would be its message.